File: map2/platform/apps/satellite-service/app/messaging/events.py
"""Domain event construction (SRS §3.2)."""
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

EVENTS_EXCHANGE = "platform.events"
EVENT_JOB_SUBMITTED = "job.submitted"
EVENT_JOB_COMPLETED = "job.completed"
EVENT_JOB_FAILED = "job.failed"


def _envelope(
    event_type: str, organisation_id: str, payload: dict[str, Any]
) -> dict[str, Any]:
    return {
        "eventId": str(uuid.uuid4()),
        "eventType": event_type,
        "occurredAt": datetime.now(timezone.utc).isoformat(),
        "organisationId": organisation_id,
        "correlationId": str(uuid.uuid4()),
        "payload": payload,
    }


def job_completed_event(
    organisation_id: str,
    job_id: str,
    job_type: str,
    output_types: list[str],
    duration_seconds: int,
) -> dict[str, Any]:
    return _envelope(
        EVENT_JOB_COMPLETED,
        organisation_id,
        {
            "jobId": job_id,
            "jobType": job_type,
            "outputTypes": output_types,
            "durationSeconds": duration_seconds,
        },
    )


def job_failed_event(
    organisation_id: str, job_id: str, job_type: str, error_message: str
) -> dict[str, Any]:
    return _envelope(
        EVENT_JOB_FAILED,
        organisation_id,
        {
            "jobId": job_id,
            "jobType": job_type,
            "errorMessage": error_message,
            "retryable": False,
            "retryCount": 0,
        },
    )

File: map2/platform/packages/shared-types/src/events.ts
/**
 * Domain event types for the RabbitMQ event bus — SRS §3.2 / FR-JOB-003.
 */
import type { ISODateString, UUID } from './common.js';
import type { JobStatus, JobType } from './job.js';
import type { OutputType } from './processing-output.js';

/**
 * RabbitMQ routing keys for durable domain events. All events are published
 * to the topic exchange {@link EVENTS_EXCHANGE}.
 */
export enum EventType {
  IMAGERY_UPLOADED = 'imagery.uploaded',
  JOB_SUBMITTED = 'job.submitted',
  JOB_STARTED = 'job.started',
  JOB_PROGRESS = 'job.progress',
  JOB_COMPLETED = 'job.completed',
  JOB_FAILED = 'job.failed',
  OUTPUT_READY = 'output.ready',
  ORGANISATION_SUSPENDED = 'organisation.suspended',
}

/** Topic exchange carrying every domain event. */
export const EVENTS_EXCHANGE = 'platform.events' as const;

/** Envelope wrapping every event published to the bus. */
export interface DomainEvent<TPayload> {
  readonly eventId: UUID;
  readonly eventType: EventType;
  readonly occurredAt: ISODateString;
  /** Tenant the event belongs to — consumers MUST scope all work to this. */
  readonly organisationId: UUID;
  /** Correlation id propagated end-to-end for tracing (SRS NFR-OBS-001/003). */
  readonly correlationId: UUID;
  readonly payload: TPayload;
}

/** Payload — imagery upload completed and checksum-verified (SRS FR-IMG-008). */
export interface ImageryUploadedPayload {
  readonly missionId: UUID;
  readonly projectId: UUID;
  readonly workspaceId: UUID;
  readonly fileCount: number;
  readonly rawSizeBytes: number;
  readonly uploadedByUserId: UUID;
}

/** Payload — a job was submitted to the processing queue. */
export interface JobSubmittedPayload {
  readonly jobId: UUID;
  readonly jobType: JobType;
  readonly missionId: UUID;
  readonly projectId: UUID;
}

/** Payload — periodic job progress tick. */
export interface JobProgressPayload {
  readonly jobId: UUID;
  readonly status: JobStatus;
  readonly percent: number;
  readonly stageLabel: string;
}

/** Payload — job finished successfully. */
export interface JobCompletedPayload {
  readonly jobId: UUID;
  readonly jobType: JobType;
  readonly outputTypes: readonly OutputType[];
  readonly durationSeconds: number;
}

/** Payload — job failed (plain-language message only; SRS FR-JOB-005). */
export interface JobFailedPayload {
  readonly jobId: UUID;
  readonly jobType: JobType;
  readonly errorMessage: string;
  readonly retryable: boolean;
  readonly retryCount: number;
}

/** Concrete, fully-typed event aliases. */
export type ImageryUploadedEvent = DomainEvent<ImageryUploadedPayload>;
export type JobSubmittedEvent = DomainEvent<JobSubmittedPayload>;
export type JobProgressEvent = DomainEvent<JobProgressPayload>;
export type JobCompletedEvent = DomainEvent<JobCompletedPayload>;
export type JobFailedEvent = DomainEvent<JobFailedPayload>;
